task3/task3_solution.py

```diff
@@ -0,0 +1,124 @@
+class Journal:
+
+    def __init__(self, students):
+        self.student_data = students
+
+    def to_letters(self, grades):
+        res = ""
+        for grade in grades:
+            if grade >= 95.0:
+                res += "A"
+            elif grade >= 85.0:
+                res += "B"
+            elif grade >= 75.0:
+                res += "C"
+            elif grade >= 65.0:
+                res += "D"
+            elif grade >= 60.0:
+                res += "E"
+            elif grade < 60.0:
+                res += "F"
+            res += " "
+        return res.strip()
+
+    # calc overall average grade
+    # not universal
+    def avg_grade(self, homework, quizzes, tests):
+        grades = homework + quizzes + tests
+        res = sum(grades)/len(grades)
+        return f"{res:.2f}"
+
+    def get_letter_grades(self, *args, throw_exception=False):
+        for dict in self.student_data:
+            if dict.get("name", 0) in args:
+                print("Grades of " + dict.get("name", 0) + ": ")
+                print("Homework:", self.to_letters(dict.get("homework", 0)))
+                print("Quizzes:", self.to_letters(dict.get("quizzes", 0)))
+                print("Tests:", self.to_letters(dict.get("tests", 0)))
+            else:
+                # if keyword specified, show error instead of skipping
+                if throw_exception:
+                    print("No student with such name in table")
+
+    def get_student_average(self, *args, throw_exception=False):
+        for dict in self.student_data:
+            if dict.get("name", 0) in args:
+                print("Average grade of", dict.get("name", 0), ":",
+                      #adding 0 if no such key found, care
+                      self.avg_grade(dict.get("homework", 0),
+                                     dict.get("quizzes", 0),
+                                     dict.get("tests", 0)))
+                # debug
+                # print(dict.get("name", 0) in (i for i in args))
+            else:
+                if throw_exception:
+                    print("No student with such name in table")
+
+    # to see the state of data
+    def show_data(self):
+        for data in self.student_data:
+            print(data.values())
+
+    def update_grade(self, student_name, throw_exception=False, **kwargs):
+        for dict in self.student_data:
+            if student_name in dict.get("name"):

```

Comment:
dict.get("name") returns None if "name" was not found, and you are looking for student_name in None
You can simply make dict.get("name")

Reply:
And dict is built in function name, it's better to change it to something more specific to that case